core-processor/src/main/java/io/micronaut/expressions/parser/ast/operator/binary/math/AddOperator.java

```diff
@@ -0,0 +1,156 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.expressions.parser.ast.operator.binary.math;
+
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.expressions.parser.ast.ExpressionNode;
+import io.micronaut.expressions.parser.ast.operator.binary.BinaryOperator;
+import io.micronaut.expressions.parser.ast.types.TypeDescriptors;
+import io.micronaut.expressions.parser.compilation.ExpressionCompilationContext;
+import io.micronaut.expressions.parser.exception.ExpressionCompilationException;
+import org.objectweb.asm.Type;
+import org.objectweb.asm.commons.GeneratorAdapter;
+import org.objectweb.asm.commons.Method;
+
+import java.util.Map;
+import java.util.Optional;
+
+import static io.micronaut.expressions.parser.ast.types.TypeDescriptors.STRING;
+import static io.micronaut.expressions.parser.ast.types.TypeDescriptors.VOID;
+import static io.micronaut.expressions.parser.ast.types.TypeDescriptors.computeNumericOperationTargetType;
+import static io.micronaut.expressions.parser.ast.types.TypeDescriptors.isNumeric;
+import static io.micronaut.expressions.parser.ast.util.EvaluatedExpressionCompilationUtils.pushPrimitiveCastIfNecessary;
+import static io.micronaut.expressions.parser.ast.util.EvaluatedExpressionCompilationUtils.pushUnboxPrimitiveIfNecessary;
+import static org.objectweb.asm.Opcodes.DADD;
+import static org.objectweb.asm.Opcodes.DUP;
+import static org.objectweb.asm.Opcodes.FADD;
+import static org.objectweb.asm.Opcodes.IADD;
+import static org.objectweb.asm.Opcodes.LADD;
+import static org.objectweb.asm.Opcodes.NEW;
+
+/**
+ * Expression node for binary '+' operator. Works both for math operation and string
+ * concatenation.
+ *
+ * @author Sergey Gavrilov
+ * @since 4.0.0
+ */
+@Internal
+public class AddOperator extends BinaryOperator {
+
+    private static final Map<String, Integer> ADD_OPERATION_OPCODES = Map.of(
+        "D", DADD,
+        "I", IADD,
+        "F", FADD,
+        "J", LADD);
+
+    private static final Type STRING_BUILDER_TYPE = Type.getType(StringBuilder.class);
+
+    private static final Method STRING_BUILD_CONSTRUCTOR =
+        new Method("<init>", VOID, new Type[]{});
+
+    private static final Method STRING_BUILD_TO_STRING =
+        new Method("toString", STRING, new Type[]{});
+
+    public AddOperator(ExpressionNode leftOperand, ExpressionNode rightOperand) {
+        super(leftOperand, rightOperand);
+    }
+
+    @Override
+    protected Type resolveOperationType(Type leftOperandType, Type rightOperandType) {
+        if (!(leftOperandType.equals(STRING)
+                  || rightOperandType.equals(STRING)
+                  || (isNumeric(leftOperandType) && isNumeric(rightOperandType)))) {
+            throw new ExpressionCompilationException(
+                "'+' operation can only be applied to numeric and string types");
```

Comment:
should we support BigInteger/BigDecimal?

Reply:
Yes, I think so. However, there are still many features described in #8323 which are not implemented as part of this PR yet. It may be preferrable to finalize this PR and add extra features in separate PRs which will only concentrate on new operations rather than all expression evaluation logic in general